File: java-ver/src/main/java/com/prvz/url_checker/validation/UrlListValidator.java
package com.prvz.url_checker.validation;

import org.springframework.util.CollectionUtils;

import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;
import java.util.List;
import java.util.stream.Collectors;

import static com.prvz.url_checker.util.Utils.isUrl;
import static java.lang.String.format;

public class UrlListValidator implements ConstraintValidator<UrlList, List<?>> {

    @Override
    public boolean isValid(List<?> value, ConstraintValidatorContext context) {

        if (CollectionUtils.isEmpty(value)) {
            return false;
        }
        List<String> incorrect = value.stream()
                .map(Object::toString)
                .filter(url -> !isUrl(url))
                .collect(Collectors.toList());
        if (incorrect.isEmpty()) {
            return true;
        }
        context.buildConstraintViolationWithTemplate(
                format(context.getDefaultConstraintMessageTemplate(),
                        String.join(",\n", incorrect)))
                .addConstraintViolation();
        return false;
    }
}

File: src/main/kotlin/com/prvz/url_checker/validation/UrlListValidator.kt
package com.prvz.url_checker.validation

import org.springframework.util.ResourceUtils
import javax.validation.ConstraintValidator
import javax.validation.ConstraintValidatorContext

open class UrlListValidator : ConstraintValidator<UrlsList, List<Any>> {

    private lateinit var urls: List<Any>

    override fun isValid(value: List<Any>?, context: ConstraintValidatorContext?): Boolean {
        return value
                ?.let {
                    if (it.isEmpty()) false
                    else {
                        it.forEach { url ->
                            if (!ResourceUtils.isUrl(url.toString())) {
                                context!!.buildConstraintViolationWithTemplate(context.defaultConstraintMessageTemplate).apply {
                                    addPropertyNode("urls")
                                    addConstraintViolation()
                                }
                                return false
                            }
                        }
                        true
                    }
                }
                ?: false
    }
}